cdc/sorter/leveldb/sorter.go

```diff
@@ -0,0 +1,262 @@
+			iterFirstSlowDuration: time.Duration(cfg.IteratorSlowReadDuration) * time.Millisecond,
+
+			metricIterFirst:   metricIterDuration.WithLabelValues("first"),
+			metricIterRelease: metricIterDuration.WithLabelValues("release"),
+		},
+		delete: deleteThrottle{
+			countThreshold: cfg.CompactionDeletionThreshold,
+			period:         time.Duration(cfg.CompactionPeriod * int(time.Second)),
+		},
+
+		lastSentResolvedTs: startTs,
+		outputCh:           outputCh,
+
+		metricIterReadDuration: metricIterDuration.WithLabelValues("read"),
+		metricIterNextDuration: metricIterDuration.WithLabelValues("next"),
+	}
+	rmb := actor.NewMailbox(actorID, sorterInputCap)
+	err = readerSystem.Spawn(rmb, r)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	return &Sorter{
+		common:        c,
+		writerRouter:  writerRouter,
+		writerActorID: actorID,
+		readerRouter:  readerRouter,
+		readerActorID: actorID,
+		outputCh:      outputCh,
+	}, nil
+}
+
+// Run runs Sorter
+func (ls *Sorter) Run(ctx context.Context) error {
+	var err error
+	select {
+	case <-ctx.Done():
+		err = ctx.Err()
+	case err = <-ls.errCh:
+	}
+	// TODO caller should pass context.
+	deadline := time.Now().Add(1 * time.Second)
+	ctx, cancel := context.WithDeadline(context.TODO(), deadline)
+	defer cancel()
+	atomic.StoreInt32(&ls.closed, 1)
+	_ = ls.writerRouter.SendB(
+		ctx, ls.writerActorID, actormsg.StopMessage())
+	_ = ls.readerRouter.SendB(
+		ctx, ls.readerActorID, actormsg.StopMessage())
+	return errors.Trace(err)
+}
+
+// AddEntry adds an RawKVEntry to the EntryGroup
+func (ls *Sorter) AddEntry(ctx context.Context, event *model.PolymorphicEvent) {
+	if atomic.LoadInt32(&ls.closed) != 0 {
+		return
+	}
+	msg := actormsg.SorterMessage(message.Task{
+		UID:        ls.uid,
+		TableID:    ls.tableID,
+		InputEvent: event,
+	})
+	_ = ls.writerRouter.SendB(ctx, ls.writerActorID, msg)
+}
+
+// TryAddEntry tries to add an RawKVEntry to the EntryGroup
+func (ls *Sorter) TryAddEntry(
+	ctx context.Context, event *model.PolymorphicEvent,
+) (bool, error) {
+	if atomic.LoadInt32(&ls.closed) != 0 {
+		return false, nil
+	}
+	msg := actormsg.SorterMessage(message.Task{
+		UID:        ls.uid,
+		TableID:    ls.tableID,
+		InputEvent: event,
+	})
+	err := ls.writerRouter.Send(ls.writerActorID, msg)
+	if err != nil {
+		if cerror.ErrMailboxFull.Equal(err) {
+			return false, nil
+		}
+		return false, errors.Trace(err)
+	}
+	return true, nil
+}
+
+// Output returns the sorted raw kv output channel
+func (ls *Sorter) Output() <-chan *model.PolymorphicEvent {
+	// Notify reader to read sorted events
+	msg := actormsg.SorterMessage(message.Task{
+		UID:     ls.uid,
+		TableID: ls.tableID,
+		ReadTs:  message.ReadTs{},
+	})
+	// It's ok to ignore error, as reader is either channel full or stopped.
+	// If it's channel full, it has been notified by others, and caller will
+	// receive new resolved events eventually.
+	_ = ls.readerRouter.Send(ls.readerActorID, msg)
```

Comment:
we call `_ = w.readerRouter.Send(w.readerActorID, msg)` in writer too,  do we need some improvement to reduce the actor messages?

Reply:
Good point! I'll add a TODO here.